presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -391,11 +390,12 @@ public void recordHeartbeat()
     @Override
     public void pruneInfo()
     {
-        QueryInfo queryInfo = finalQueryInfo.get();
-        if (queryInfo == null || !queryInfo.getOutputStage().isPresent()) {
+        Optional<QueryInfo> finalQueryInfo = stateMachine.getFinalQueryInfo();
```

Comment:
Can we move this code inside of `QueryStateMachine` so we don't have expose the `setFinalQueryInfo` at all?